README.md
# AC KeyLock
This tool can be used to lock operations (e.g. make sure workers will fetch a job only once). You can use it with Redis as distributed lock systems or with the built-in memory cache.

This package is a successor of ac-redisLock and can easily replace ac-redisLock. Please note, that callback-style is no longer available.

## Usage
### Init
Init function requires a valid redis instance.
+ redis - OPTIONAL INSTANCE ioredis instance
+ logger - OPTIONAL INSTANCE logger instance (e.g. Winston)
+ logLevel - OPTIONAL STRING logLevel for the Redis operation

```
await keylock.init()
// OR WITH REDIS
await keylock.init({ redis: REDISINSTANCE })
```

### Lock key
LockKey has the following parameters
+ key - STRING required name for the key
+ expires - OPTIONAL INT seconds after the lock is released, defaults to 10 seconds
+ value - OPTIONAL STRING value for this redisKey. Use it to compare/secure lock values

```
await keylock.lockKey(params)
// return values
{ status: 423 } -> there is already a lock on this key
{ value: 'xxxx' } -> a lock has been put on this key
```

### Release lock
ReleaseKey has the following parameters
+ key - STRING required name for the key
+ value - OPTIONAL STRING if set will be compared with the redisKey value before releasing. If not matching, the function will return an error message

```
await keylock.releaseLock(params, [callback])
// return values
{ status: 200 } -> lock has been released and deleted
```


## Examples

### Usage with Redis
```
const keyLock = require('ac-keylock')

// create a Redis instance (using io-redis)
const app.redisInstance = new Redis(options)

await keyLock.init({
  redis: app.redisInstance 
})

const params = {
  key: 'someMeaningfulKey',
}
let result = await keyLock.lockKey(params)
// { status: 423 } -> key is locked
// { value } -> value for this lock -> set as params.value for releaseLock

// redisKey is deleted and lock released
await keyLock.releaseKey(params)
// { status: 200 }
```

### Usage with built-in memory cache
```
const keyLock = require('ac-keylock')

await keyLock.init()

const params = {
  key: 'someMeaningfulKey',
}
let result = await keyLock.lockKey(params)
// { status: 423 } -> key is locked
// { value } -> value for this lock -> set as params.value for releaseLock

// redisKey is deleted and lock released
await keyLock.releaseKey(params)
// { status: 200 }

```

## Links
- [Website](https://www.admiralcloud.com/)
- [Twitter (@admiralcloud)](https://twitter.com/admiralcloud)
- [Facebook](https://www.facebook.com/MediaAssetManagement/)

## License
[MIT License](https://opensource.org/licenses/MIT) Copyright © 2009-present, AdmiralCloud AG, Mark Poepping

CHANGELOG.md
<a name="0.0.3"></a>

## [0.0.3](https://github.com/admiralcloud/ac-keylock/compare/v0.0.2..v0.0.3) (2021-05-01 06:40:22)


### Bug Fix

* **App:** Replace state with status | MP | [c865e9559c85e5c67dc877ccfa95537e9ef92b6b](https://github.com/admiralcloud/ac-keylock/commit/c865e9559c85e5c67dc877ccfa95537e9ef92b6b)    
Use status (same as HTTP status code) instead of state.
<a name="0.0.2"></a>

## [0.0.2](https://github.com/admiralcloud/ac-keylock/compare/v0.0.1..v0.0.2) (2021-05-01 06:33:44)


### Bug Fix

* **App:** Use global var instead of this | MP | [91f253bb28013e29c559551ca7e346049a2e64f6](https://github.com/admiralcloud/ac-keylock/commit/91f253bb28013e29c559551ca7e346049a2e64f6)    
Use global var instead of this
### Documentation

* **App:** Typo fix | MP | [02d562a97a3a996b2e54871138e43e76bea5f84c](https://github.com/admiralcloud/ac-keylock/commit/02d562a97a3a996b2e54871138e43e76bea5f84c)    
Typo fix
### Chores

* **App:** Updated .gitignore | MP | [044d031b73a38b8aad2ac30a61ea7a2ae25bc9b3](https://github.com/admiralcloud/ac-keylock/commit/044d031b73a38b8aad2ac30a61ea7a2ae25bc9b3)    
Ignore .npmrc
<a name="0.0.1"></a>

## [0.0.1](https://github.com/admiralcloud/ac-keylock/compare/..v0.0.1) (2021-04-12 18:48:10)


### Bug Fix

* **App:** Initial version | MP | [a4eb9fef4ddca5b220db77dbbb29b156421d9ce2](https://github.com/admiralcloud/ac-keylock/commit/a4eb9fef4ddca5b220db77dbbb29b156421d9ce2)    
Initial version


test/test.js
const keyLock = require('../index')
const expect = require('chai').expect
const { v4: uuidV4 } = require('uuid')

const Redis = require('ioredis')
const redis = new Redis()

const testkey = uuidV4()
const expectedLockValue = uuidV4()
let lockValue

const useRedis = function() {
  describe('Use with Redis',  function() {
    this.timeout(5000)
  
    before(async() => {
      await keyLock.init({
        redis
      })
    })
  
    after(() => {
      redis.quit()
    })
  
    describe('Run test', () => {
      it('Lock for 2 seconds', async() => {
        const params = {
          key: testkey,
          value: expectedLockValue,
          expires: 2
        }
        let r = await keyLock.lockKey(params)
        expect(r).to.have.property('value', expectedLockValue)
      })
  
      it('Lock again - should fail - key is locked', async() => {
        const params = {
          key: testkey,
          expires: 2
        }
        let r = await keyLock.lockKey(params)
        expect(r).to.have.property('status', 423)
      })
  
      it('Wait for 2 seconds', done => {
        setTimeout(done, 2000)
      })
  
      it('Lock for 5 seconds - should work again', async() => {
        const params = {
          key: testkey,
          value: expectedLockValue,
          expires: 2
        }
        let r = await keyLock.lockKey(params)
        expect(r).to.have.property('value', expectedLockValue)
        lockValue = r.value
      })
  
      it('Release lock', async() => {
        const params = {
          key: testkey,
          lockValue
        }
        let r = await keyLock.releaseLock(params)
        expect(r).to.have.property('status', 200)
      })
    })
  })
}

const useMemory = function() {
  describe('Use with Memory',  function() {
    this.timeout(5000)
  
    before(async() => {
      await keyLock.init({})
    })
  
  
    describe('Run test', () => {
      it('Lock for 2 seconds', async() => {
        const params = {
          key: testkey,
          value: expectedLockValue,
          expires: 2
        }
        let r = await keyLock.lockKey(params)
        expect(r).to.have.property('value', expectedLockValue)
      })
  
      it('Lock again - should fail - key is locked', async() => {
        const params = {
          key: testkey,
          expires: 2
        }
        let r = await keyLock.lockKey(params)
        expect(r).to.have.property('status', 423)
      })
  
      it('Wait for 2 seconds', done => {
        setTimeout(done, 2000)
      })
  
      it('Lock for 5 seconds - should work again', async() => {
        const params = {
          key: testkey,
          value: expectedLockValue,
          expires: 2
        }
        let r = await keyLock.lockKey(params)
        expect(r).to.have.property('value', expectedLockValue)
        lockValue = r.value
      })
  
      it('Release lock', async() => {
        const params = {
          key: testkey,
          value: lockValue
        }
        let r = await keyLock.releaseLock(params)
        expect(r).to.have.property('status', 200)
      })
    })
  })
}

const handleErrors = function() {

  describe('Handle errors with Memory',  function() {
    this.timeout(5000)
  
    before(async() => {
      await keyLock.init({})
    })

    describe('Run test', () => {
      it('Missing key', async() => {
        const params = {
          value: expectedLockValue,
          expires: 2
        }
        try {
          await keyLock.lockKey(params)
        }
        catch (e) {
          expect(e).to.be.instanceOf(Error)
          expect(e.message).to.eql('ackeylock_lockKey_key_isRequired')
        }
      })

      it('Lock for 2 seconds', async() => {
        const params = {
          key: testkey,
          value: expectedLockValue,
          expires: 2
        }
        let r = await keyLock.lockKey(params)
        expect(r).to.have.property('value', expectedLockValue)
        lockValue = r.value
      })

      it('Release lock with wrong value - should fail', async() => {
        const params = {
          key: testkey,
          value: 'abc'
        }
        try {
          await keyLock.releaseLock(params)
        }
        catch (e) {
          expect(e).to.be.instanceOf(Error)
          expect(e.message).to.eql('ackeylock_releaseLock_valueMismatch')
        }
      })

      it('Release lock with value - should work', async() => {
        const params = {
          key: testkey,
          value: lockValue
        }
        const r = await keyLock.releaseLock(params)
        expect(r).to.have.property('status', 200)

      })

    })
  })
}


// test suites
useRedis()
useMemory()
handleErrors()

index.js
const _ = require('lodash')
const { v4: uuidV4 } = require('uuid')

const NodeCache = require( "node-cache" )

const keyLock = () => {

  let lockBaseParams = {} 
 
  /**
   * @param params.redis {Instance} Redis instance to use
   * @param params.logger {Instance} optional logger (e.g. Winston). Falls back to console
   * @param params.logLevel {String} optional logLevel
   */

  const init = async (params) => {
    lockBaseParams = {
      redis: _.get(params, 'redis'),
      logger: _.get(params, 'logger', console),
      logLevel: _.get(params, 'logLevel', 'log')
    }

    // make a test connection
    if (lockBaseParams.redis) {
      const testKey = uuidV4()
      try {
        await lockBaseParams.redis.set(testKey, 1, 'EX', 1)
      }
      catch(e) {
        lockBaseParams.logger['error']({ message: 'cannotConnectToRedis' })
      }
    }
    else {
      // use memory cache
      lockBaseParams.storage = new NodeCache()
    }
  }

  /**
   * @param params.key {String} REQUIRED identifier
   * @param params.expires {Integer} optional seconds to expire the key automatically
   * @param cb (err or null) err can be 423 (redis key is locked) or a real error or null
   */
  const lockKey = async (params) => {
    const key = params.key
    if (!key) {
      throw Error('ackeylock_lockKey_key_isRequired')
    }
    const expires = (params.expires && parseInt(params.expires)) || 10 // 10 seconds default value
    const value = params.value || uuidV4()

    if (lockBaseParams.redis) {
      let result = await lockBaseParams.redis.set(key, value, 'EX', expires, 'NX')
      if (result === 'OK') return { value }
      else return { status: 423 }
    }
    else {
      if (lockBaseParams.storage.has(key)) return { status: 423 }
      lockBaseParams.storage.set(key, value, expires)
      return { value }
    }
  }

  const releaseLock = async (params) => {
    const key = params.key
    if (!key) {
      throw Error('ackeylock_releaseLock_key_isRequired')
    }
    const value = params.value

    if (lockBaseParams.redis) {
      if (value) {
        let result = await lockBaseParams.redis.get(key)
        if (result !== value) {
          lockBaseParams.logger['error']('ackeylock_releaseLock_valueMismatch | %j', { key, expected: result, value })
          throw Error('ackeylock_releaseLock_valueMismatch')
        }  
      }
      await lockBaseParams.redis.del(key)
      return { status: 200 }
    }
    else {
      if (value) {
        let result = lockBaseParams.storage.get(key)
        if (result !== value) {
          lockBaseParams.logger['error']('ackeylock_releaseLock_valueMismatch | %j', { key, expected: result, value })
          throw Error('ackeylock_releaseLock_valueMismatch')  
        }
      }
      lockBaseParams.storage.del(key)
      return { status: 200 }
    }
  }

  return {
    init,
    lockKey,
    releaseLock
  }
}

module.exports = keyLock()
